pkg/virt-controller/watch/vm.go

```diff
@@ -83,222 +126,433 @@ func (c *VMController) runWorker() {
-		// unique VM instance, delete them before continuing ...
-		if len(pods.Items) > 0 && hasPod == false {
-			logger.Error("VM Pods do already exist, will clean up before continuing.")
-			if err := c.vmService.DeleteVMPod(vmCopy); err != nil {
-				logger.Reason(err).Error("Deleting VM pods failed.")
-				return err
-			}
-			// the pod informer will reenqueue the key as a result of it being deleted.
-			return nil
-		}
+	vmCopy := vm.DeepCopy()
 
-		// Defaulting and setting constants
-		// TODO move defaulting to virt-api
-		kubev1.SetObjectDefaults_VirtualMachine(vmCopy)
+	switch {
 
-		if hasPod == false {
-			// Create a Pod which will be the VM destination
-			if err := c.vmService.StartVMPod(vmCopy); err != nil {
-				logger.Reason(err).Error("Defining a target pod for the VM failed.")
-				return err
+	case vm.IsUnprocessed():
+		if podExists {
+			vmCopy.Status.Phase = virtv1.Scheduling
+		} else if vm.DeletionTimestamp != nil {
+			vmCopy.Status.Phase = virtv1.Failed
+		} else {
+			vmCopy.Status.Phase = virtv1.Pending
+		}
+	case vm.IsScheduling():
+		switch {
+		case podExists:
+			if isPodOwnedByHandler(pod) {
+				// vm is still owned by the controller but pod is already handed over,
+				// so let's hand over the vm too
+				vmCopy.Status.Interfaces = []virtv1.VirtualMachineNetworkInterface{
+					{
+						IP: pod.Status.PodIP,
+					},
+				}
+				vmCopy.Status.Phase = virtv1.Scheduled
+				if vmCopy.Labels == nil {
+					vmCopy.Labels = map[string]string{}
+				}
+				vmCopy.ObjectMeta.Labels[virtv1.NodeNameLabel] = pod.Spec.NodeName
+				vmCopy.Status.NodeName = pod.Spec.NodeName
+			} else if isPodDownOrGoingDown(pod) {
+				vmCopy.Status.Phase = virtv1.Failed
 			}
+		case !podExists:
+			// someone other than the controller deleted the pod unexpectedly
+			vmCopy.Status.Phase = virtv1.Failed
 		}
-
-		// Mark the VM as "initialized". After the created Pod above is scheduled by
-		// kubernetes, virt-handler can take over.
-		vmCopy.Status.Phase = kubev1.Scheduling
-		if _, err := c.clientset.VM(vm.Namespace).Update(vmCopy); err != nil {
-			logger.Reason(err).Error("Updating the VM state to 'Scheduling' failed.")
-			return err
+	case vm.IsFinal():
+		if !podExists {
+			controller.RemoveFinalizer(vmCopy, virtv1.VirtualMachineFinalizer)
 		}
-		logger.Info("Handing over the VM to the scheduler succeeded.")
-	case kubev1.Scheduling:
-		// Target Pod for the VM was already created, check if it is  running and update the VM to Scheduled
+	case vm.IsRunning() || vm.IsScheduled():
+		// Don't process states where the vm is clearly owned by virt-handler
+		return nil
+	default:
+		return fmt.Errorf("unknown vm phase %v", vm.Status.Phase)
+	}
 
-		// Deep copy the object, so that we can safely manipulate it
-		vmCopy := vm.DeepCopy()
-		logger = log.Log.Object(vmCopy)
+	// Select the right failure reason in case we have an error
+	reason := ""
+	if !podExists {
```

Comment:
We shouldn't have to guess the reason for the sync failure.

Reply:
Done